activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -21,7 +21,9 @@ def dirties_query_cache(base, *method_names)
           method_names.each do |method_name|
             base.class_eval <<-end_code, __FILE__, __LINE__ + 1
               def #{method_name}(...)
-                ActiveRecord::Base.clear_query_caches_for_current_thread
+                if query_cache_enabled || pool.query_cache_enabled
```

Comment:
```suggestion
                if pool.query_cache_enabled?
```

I think we only care about the pool. Because we might be doing queries on a connection that generally have a query cache and might make it invalid, but have it temporarily disabled.

Reply:
There are about 12 test cases that fail without the connection check (e.g. https://github.com/rails/rails/blob/cff2eb1be11cb729382b3674e3caa2bd198ec55e/activerecord/test/cases/query_cache_test.rb#L890-L898).

I looked like something that would be a undesirable change. For example if you've got a script that isn't run through "rails runner" and you use `ActiveRecord::Base.cache` in it, the query cache will be used but not cleared.

Reply:
Uh, I'm confused here. 

Previously we'd always clear it, so I don't see how removing that condition would change something. Does `pool.query_cache_enabled` properly default to `true`?

Reply:
Also `ActiveRecord::ConnectionAdapters::ConnectionPool` doesn't have a `query_cache_enabled` method. I can alias `query_cache_enabled` to it though if we want. We'd probably want to do the same to the connection for consistency.

Reply:
> Also `ActiveRecord::ConnectionAdapters::ConnectionPool` doesn't have a `query_cache_enabled?` method.

Aaah! my bad, I missed this method was pre-existing.

Yeah, what I had in mind was a pool configuration, but this isn't a config, it's a state, so yeah we can't use that.


I wonder if we should simply look at fixing the problem of clearing all caches in multi-db situations. There must be an heuristic we can use to scope the clearing. e.g. use the `connection_specification_name`, or DB name or something.

Reply:
> Previously we'd always clear it, so I don't see how removing that condition would change something. Does pool.query_cache_enabled properly default to true?

It depends on whether you are wrapped in a Rails executor or not. Outside it no, but inside one, it will have been enabled.

Runner this script from within a rails app (with ruby directly, not via the rails runner) - you can see how to get all the combinations of connection and pool enabled/disabled settings:

```ruby
require_relative "config/environment"

def pool_and_connection_enabled
  p [ActiveRecord::Base.connection_pool.query_cache_enabled, ActiveRecord::Base.connection.query_cache_enabled]
end

pool_and_connection_enabled

Rails.application.executor.wrap {
  pool_and_connection_enabled
}

Rails.application.executor.wrap {
  ActiveRecord::Base.connection_pool.disable_query_cache!

  pool_and_connection_enabled
}

Rails.application.executor.wrap {
  ActiveRecord::Base.connection_pool.disable_query_cache!

  ActiveRecord::Base.cache do
    pool_and_connection_enabled
  end
}
```

Output:
```
[false, false]
[true, true]
[false, false]
[false, true]
```

> I wonder if we should simply look at fixing the problem of clearing all caches in multi-db situations. There must be an heuristic we can use to scope the clearing. e.g. use the connection_specification_name, or DB name or something.

There was [an attempt at this before](https://github.com/rails/rails/pull/40622) to only clear the current pool's connections, but [it was reverted](https://github.com/rails/rails/pull/45259).

Reply:
> There was [an attempt at this before](https://github.com/rails/rails/pull/40622) to only clear peer connections, but [it was reverted](https://github.com/rails/rails/pull/45259).

Right, but it was based on a lot of unknown, might be time to revisit it.

I understand you are in the need for a fix or workaround, but I'm thinking it would be best to improve the situation rather than to add one more ad hoc solution (sorry for not noticing all this immediately and sending you on goose chases).

Reply:
> I understand you are in the need for a fix or workaround, but I'm thinking it would be best to improve the situation rather than to add one more ad hoc solution (sorry for not noticing all this immediately and sending you on goose chases).

No that's fine, there's no point adding more things on top of what's there to get the behaviour we want. We might find that whatever we do we'll need to make the cross-pool invalidation behaviour configurable, but it makes sense to get it to work the way we want first and then add configuration options if necessary afterwards.